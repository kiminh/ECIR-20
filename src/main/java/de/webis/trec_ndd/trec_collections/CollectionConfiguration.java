package de.webis.trec_ndd.trec_collections;

import java.io.InputStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;

public interface CollectionConfiguration extends Serializable {

	public String getPathToCollection();
	public String getCollectionType();
	public String getDocumentGenerator();
	public List<SharedTask> getSharedTasks();

	@SneakyThrows
	public default Set<String> judgedDocumentIds() {
		Set<String> ret = new HashSet<>();
		
		for(String qrelResource : getQrelResourcesOfAllSharedTasks()) {
			InputStream resource = AnseriniCollectionReader.class.getResourceAsStream(qrelResource);
			ret.addAll(new QrelReader().readJudgedDocumentIdsFromQrels(resource));
		}
		
		return ret;
	}
	
	public default HashSet<String> documentIdsInRunFiles() {
		HashSet<String> ret = new HashSet<>();
		
		for(SharedTask sharedTask: getSharedTasks()) {
			ret.addAll(sharedTask.documentIdsInRunFiles());
		}
		
		return ret;
	}
	
	public default List<String> getQrelResourcesOfAllSharedTasks() {
		return getSharedTasks().stream()
				.map(SharedTask::getQrelResource)
				.collect(Collectors.toList());
	}
	
	@Getter
	@AllArgsConstructor
	public static enum TrecCollections implements CollectionConfiguration {
		CLUEWEB09(
			// compressed 4TB
			"/mnt/ceph/storage/corpora/corpora-thirdparty/corpus-clueweb09/",
			"ClueWeb09Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
				TrecSharedTask.WEB_2009,
				TrecSharedTask.WEB_2010,
				TrecSharedTask.WEB_2011,
				TrecSharedTask.WEB_2012,
				TrecSharedTask.SESSION_2010,
				TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012
			))
		),
		CLUEWEB12(
			"/mnt/ceph/storage/data-in-progress/kibi9872/clueweb12/",
			"ClueWeb12Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.WEB_2013,
					TrecSharedTask.WEB_2014,
					TrecSharedTask.SESSION_2013,
					TrecSharedTask.SESSION_2014
			))
		),
		GOV1(
			// compressed 4.6 GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Collections.emptyList())
		),
		GOV2(
			// compressed 81GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.TERABYTE_2004,
					TrecSharedTask.TERABYTE_2005_ADHOC,
					TrecSharedTask.TERABYTE_2006
			))
		),
		GOV2_MQ(			// compressed 81GB
				"/mnt/ceph/storage/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
				"TrecwebCollection",
				"JsoupGenerator",
				Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.MILLION_QUERY_2007,
					TrecSharedTask.MILLION_QUERY_2008
				))
		),
		ROBUST04(
			"/mnt/ceph/storage/data-in-progress/kibi9872/robust-04/",
			"TrecCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.ROBUST_04))
		),
		CORE2018(
			"/mnt/nfs/webis20/data-in-progress/trec-2018-webis/WashingtonPost/WashingtonPost.v2/data/",
			"WashingtonPostCollection",
			"WapoGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2018))
		),
		CORE2017(
			"/mnt/nfs/webis20/data-in-progress/trec-2017-webis/data/nyt/",
			"NewYorkTimesCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2017))
		);
		

		private final String pathToCollection,
							collectionType,
							documentGenerator;

		private final List<SharedTask> sharedTasks;
	}
	
	@Getter
	@AllArgsConstructor
	public static enum OtherCollections implements CollectionConfiguration {
		COMMON_CRAWL_2015_02("CommonCrawlCollection", "JsoupGenerator");
		
		private final String collectionType,
			documentGenerator;

		@Override
		public List<SharedTask> getSharedTasks() {
			return Collections.emptyList();
		}

		@Override
		public String getPathToCollection() {
			return null;
		}
	}
}
