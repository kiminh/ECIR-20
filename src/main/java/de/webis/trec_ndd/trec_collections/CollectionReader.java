package de.webis.trec_ndd.trec_collections;

import java.util.List;

public interface CollectionReader {
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection();
	
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection();

	public List<CollectionDocument> extractRawDocumentsFromCollection();
}
