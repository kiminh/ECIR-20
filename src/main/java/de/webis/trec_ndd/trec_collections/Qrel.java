package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class Qrel implements Serializable{
    private int score;
    private int topicNumber;
    private String documentID;
    
	public Qrel(String line) {
        String[] args = line.split("[\\s\\t]+");
        
        if (args.length == 5) {
        	//prels, like: 1 clueweb09-en0003-55-31884 0 0 1
            topicNumber = Integer.parseInt(args[0]);
            documentID = args[1];
            //FIXME: what is the score for prels?
            score = Integer.MIN_VALUE;
        }
        else {
            topicNumber = Integer.parseInt(args[0]);
            //skipping 1 because it is irrelevant -> has always 0 as value wehen printed
            documentID = args[2];
            score = Integer.parseInt(args[3]);	
        }
    }
	

	
	public String toString() {
		return new StringBuilder().append(topicNumber).append(" 0 ").append(documentID).append(' ').append(score).toString();
	}
	
	@Override
	public int hashCode() {
		return toString().hashCode();
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel) {
			return this.documentID.equals(((Qrel) o).documentID) && this.score==((Qrel)o).score && this.topicNumber==((Qrel)o).topicNumber;
		}
		else return false;
	}
}

