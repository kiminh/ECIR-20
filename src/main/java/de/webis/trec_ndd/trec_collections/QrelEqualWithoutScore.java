package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;

import lombok.experimental.Accessors;

@Accessors(chain = true)
@SuppressWarnings("serial")
public class QrelEqualWithoutScore extends Qrel implements Serializable{

	public QrelEqualWithoutScore() {
		
	}
	
	public QrelEqualWithoutScore(String line) {
		super(line);
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Qrel || o instanceof QrelEqualWithoutScore) {
			return this.getDocumentID().equals(((Qrel) o).getDocumentID()) && this.getTopicNumber()==((Qrel)o).getTopicNumber();
		}
		else return false;
	}
	
	@Override
	public int hashCode() {
		return new StringBuilder().append(getTopicNumber()).append(getDocumentID()).toString().hashCode();
	}
	
	@Override
	public String toString() {
		return super.toString();
	}
	
	public QrelEqualWithoutScore copy() {
		QrelEqualWithoutScore ret = new QrelEqualWithoutScore();
		ret.setDocumentID(this.getDocumentID());
		ret.setTopicNumber(this.getTopicNumber());
		ret.setScore(this.getScore());
		
		return ret;
	}
	
}