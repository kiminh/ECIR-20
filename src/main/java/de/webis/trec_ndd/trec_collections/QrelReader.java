package de.webis.trec_ndd.trec_collections;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.util.Set;

import java.util.List;
import java.util.ArrayList;
import java.util.HashSet;

public class QrelReader {

	public List<Qrel> readFile(String filepath) throws IOException{
		BufferedReader reader = new BufferedReader(new FileReader(filepath));
		String line;
		List<Qrel> results = new ArrayList<Qrel>();
		while((line=reader.readLine())!=null) {
			results.add(new Qrel(line));
		}
		reader.close();
		return results;
	}
	
	public List<QrelEqualWithoutScore> readFileQrelEqualWithoutScore(String filePath) throws IOException{
		return readFileQrelEqualWithoutScore(new File(filePath));
	}
	
	public List<QrelEqualWithoutScore> readFileQrelEqualWithoutScore(File file) throws IOException{
		BufferedReader reader = new BufferedReader(new FileReader(file));
		String line;
		List<QrelEqualWithoutScore> results = new ArrayList<QrelEqualWithoutScore>();
		while((line=reader.readLine())!=null) {
			while(line.contains("  ")) line=line.replaceAll("  "," ");
			results.add(new QrelEqualWithoutScore(line));
		}
		reader.close();
		return results;
		
	}

	public Set<String> readJudgedDocumentIdsFromQrels(File file) throws IOException {
		return readJudgedDocumentIdsFromQrels(new FileInputStream(file));
	}
	
	public Set<String> readJudgedDocumentIdsFromQrels(InputStream inputStream) throws IOException {
		BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
		String line;
		Set<String> results = new HashSet<String>();
		while((line=reader.readLine())!=null) {
			results.add(new Qrel(line).getDocumentID());
		}
		reader.close();
		return results;

		/*return new QueryJudgments(qrelsFile)
				.getQrels().values().stream()
				.map(Map::keySet)
				.flatMap(Set::stream)
				.collect(Collectors.toSet());
		*/
	}
}
