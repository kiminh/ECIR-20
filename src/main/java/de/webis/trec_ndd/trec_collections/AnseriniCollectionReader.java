package de.webis.trec_ndd.trec_collections;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

import org.apache.lucene.document.Document;

import com.google.common.collect.Iterators;

import de.webis.trec_ndd.spark.DocumentHash;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.OtherCollections;
import io.anserini.collection.ClueWeb09Collection;
import io.anserini.collection.ClueWeb12Collection;
import io.anserini.collection.CommonCrawlCollection;
import io.anserini.collection.DocumentCollection;
import io.anserini.collection.Segment;
import io.anserini.collection.SegmentProvider;
import io.anserini.collection.SourceDocument;
import io.anserini.index.IndexCollection;
import io.anserini.index.IndexCollection.Args;
import io.anserini.index.IndexCollection.Counters;
import io.anserini.index.generator.LuceneDocumentGenerator;
import lombok.Data;
import lombok.SneakyThrows;

@Data
@SuppressWarnings("serial")
public class AnseriniCollectionReader<T extends SourceDocument> implements CollectionReader, Serializable {
	private final CollectionConfiguration config;
	
	@Override
	public List<CollectionDocument> extractJudgedRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			judgedCollectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}
	
	@Override
	public List<CollectionDocument> extractRunFileDocumentsFromsCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			documentsFromRunFilesInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@Override
	public List<CollectionDocument> extractRawDocumentsFromCollection() {
		List<CollectionDocument> ret = new LinkedList<>();

		for(String segmentPath : segmentPaths()) {
			collectionDocumentsInPath(segmentPath).forEachRemaining(ret::add);
		}

		return ret;
	}

	@SneakyThrows
	public List<String> segmentPaths() {
		return documentCollection().getFileSegmentPaths()
				.stream()
				.map(p -> p.toAbsolutePath().toString())
				.collect(Collectors.toList());
	}
	
	public Iterator<DocumentHash> documentHashIterator(String segmentPath) {
		Iterator<CollectionDocument> iter = collectionDocumentsInPath(segmentPath);
		
		return Iterators.transform(iter, d -> new DocumentHash(d));
	}
	
	public Iterator<CollectionDocument> judgedCollectionDocumentsInPath(String segmentPath) {
		Set<String> judgedDocumentIds = config.judgedDocumentIds();
		
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> judgedDocumentIds.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath) {
		Set<String> documentIdsInRunFiles = config.documentIdsInRunFiles();
		
		return documentsFromRunFilesInPath(segmentPath, documentIdsInRunFiles);
	}
	
	public Iterator<CollectionDocument> documentsFromRunFilesInPath(String segmentPath, Set<String> documentIdsInRunFiles) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, doc -> documentIdsInRunFiles.contains(documentId(doc))), i -> i != null);
	}
	
	public Iterator<CollectionDocument> collectionDocumentsInPath(String segmentPath) {
		return Iterators.filter(iteratorForCollectionDocumentsInPath(segmentPath, null), i -> i != null);
	}
	
	@SneakyThrows
	private Iterator<CollectionDocument> iteratorForCollectionDocumentsInPath(String segmentPath, Function<Document, Boolean> keepDocument) {
		SegmentProvider<T> collection = documentCollection();
		
		return new AnseriniCollectionSegmentIterator<>(
				collection, 
				documentGenerator(),
				collection.createFileSegment(Paths.get(segmentPath)),
				segmentPath,
				keepDocument
		);
	}

	@Data
	private static class AnseriniCollectionSegmentIterator<T extends SourceDocument>
			implements Iterator<CollectionDocument> {
		private final SegmentProvider<T> collection;
		private final LuceneDocumentGenerator<T> generator;
		private final Segment<T> segment;
		private final String segmentPath;
		private final Function<Document, Boolean> keepDocument;

		@Override
		public boolean hasNext() {
			return segment.hasNext();
		}

		@Override
		public CollectionDocument next() {
			T document = segment.next();
			if (document == null || !document.indexable()) {
				return null;
			}
			
			Document doc = generator.createDocument(document);
			if (doc == null || (keepDocument != null && Boolean.FALSE.equals(keepDocument.apply(doc)))) {
				return null;
			} else {
				CollectionDocument ret = CollectionDocument.fromLuceneDocument(doc);
				ret.setUrl(extractUrlIfPossible(document));
				return ret;
			}
		}
	}
	
	private static URL extractUrlIfPossible(SourceDocument document) {
		try {
			return extractUrlOrFail(document);
		} catch (Exception e) {
			return null;
		}
	}

	private static URL extractUrlOrFail(SourceDocument document) throws Exception {
		if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb09Collection.Document) document).getURL());
		} else if (document instanceof ClueWeb09Collection.Document) {
			return new URL(((ClueWeb12Collection.Document) document).getURL());
		}
		
		return null;
	}
	
	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private SegmentProvider<T> documentCollection()
			throws InstantiationException, IllegalAccessException, ClassNotFoundException {
		SegmentProvider<T> ret = (SegmentProvider) Class.forName("io.anserini.collection." + args().collectionClass)
				.newInstance();
		((DocumentCollection) ret).setCollectionPath(Paths.get(config.getPathToCollection()));

		if(ret instanceof CommonCrawlCollection) {
			((CommonCrawlCollection) ret).setS3Files(((OtherCollections) config).getS3Files());
		}
		
		return ret;
	}

	@SneakyThrows
	@SuppressWarnings({ "rawtypes", "unchecked" })
	private LuceneDocumentGenerator<T> documentGenerator() throws InstantiationException, IllegalAccessException,
			IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, Exception {
		Args args = args();
		Class<?> clazz = Class.forName("io.anserini.index.generator." + args.generatorClass);

		return (LuceneDocumentGenerator) clazz.getDeclaredConstructor(Args.class, Counters.class).newInstance(args,
				new IndexCollection(args).new Counters());
	}

	private Args args() {
		Args ret = new Args();
		ret.generatorClass = config.getDocumentGenerator();
		ret.collectionClass = config.getCollectionType();
		ret.index = "-index";
		ret.input = config.getPathToCollection();

		return ret;
	}

	static String documentId(Document doc) {
		return doc.getField(LuceneDocumentGenerator.FIELD_ID).stringValue();
	}
}
