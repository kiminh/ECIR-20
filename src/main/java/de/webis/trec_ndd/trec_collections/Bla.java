package de.webis.trec_ndd.trec_collections;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.codehaus.jackson.map.ObjectMapper;

import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import lombok.Data;

public class Bla {
	@Data
	public static class BLA {
		private final String topicNumber;
		private final String topicQuery;
		private final Map<String, String> documentToJudgment; 
	}
	
	public static void main(String [] args) throws Exception {
		Map<String, Object> ret = new LinkedHashMap<>();
		
		for(SharedTask task : TrecCollections.CLUEWEB09.getSharedTasks()) {
			ret.put(task.toString(), taskStuff(task));
		}
		
		new ObjectMapper().writeValue(new File("clueweb.json"), ret);
	}
	
	private static Map<String, BLA> taskStuff(SharedTask task) {
		Map<String, BLA> ret = new HashMap<>();
		
		for(Map.Entry<String, Map<String, String>> i : task.documentJudgments().getData().entrySet()) {
			String query = task.getQueryForTopic(i.getKey());
			ret.put(i.getKey(), new BLA(i.getKey(), query, i.getValue()));
		}
		
		return ret;
	}
}
