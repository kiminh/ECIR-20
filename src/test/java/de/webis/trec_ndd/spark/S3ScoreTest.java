package de.webis.trec_ndd.spark;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.webis.trec_ndd.spark.S3ScoreOnWord8GrammIndex.S3Score;
import de.webis.trec_ndd.spark.S3ScoreOnWord8GrammIndex.S3ScoreIntermediateResult;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.generator.LuceneDocumentGenerator;

import org.apache.lucene.document.Document;
import org.apache.lucene.document.StoredField;

public class S3ScoreTest {
	@Test
	public void testS3ScoreCalculation() {
		S3ScoreIntermediateResult intermediateS3 = new S3ScoreIntermediateResult();
		intermediateS3.setCommonNGramms(1);
		intermediateS3.setLeftMetadata(docHash(null, 2));
		intermediateS3.setRightMetadata(docHash(null, 3));
		
		Approvals.verifyAsJson(new S3Score(intermediateS3));
	}
	
	@Test
	public void testS3ScoreCalculationOnShortDocuments() {
		S3ScoreIntermediateResult intermediateS3 = new S3ScoreIntermediateResult();
		intermediateS3.setCommonNGramms(0);
		intermediateS3.setLeftMetadata(docHash("a", 0));
		intermediateS3.setRightMetadata(docHash("b", 0));
		
		S3Score s3 = new S3Score(intermediateS3);
		if(Double.isNaN(s3.getS3Score())) {
			s3.setS3Score(0.0);
		}
		
		Approvals.verifyAsJson(s3);
	}
	
	@Test
	public void testS3ScoreCalculationOnShortDocumentsWithIdenticalHash() {
		S3ScoreIntermediateResult intermediateS3 = new S3ScoreIntermediateResult();
		intermediateS3.setCommonNGramms(0);
		intermediateS3.setLeftMetadata(docHash("a", 0));
		intermediateS3.setRightMetadata(docHash("a", 0));
		
		S3Score s3 = new S3Score(intermediateS3);
		Approvals.verifyAsJson(s3);
	}
	
	@Test
	public void testS3ScoreOnRealWorldExampleWithIdenticalHash() {
		CollectionDocument a = doc("Ohio Department Of Natural Resources http://www.dnr.state.oh.us", 1);
		CollectionDocument b = doc("Ohio Department Of Natural Resources http://www.dnr.state.oh.us", 2);
		
		S3ScoreIntermediateResult intermediateS3 = new S3ScoreIntermediateResult();
		intermediateS3.setCommonNGramms(0);
		intermediateS3.setLeftMetadata(new DocumentHash(a));
		intermediateS3.setRightMetadata(new DocumentHash(b));
		
		S3Score s3 = new S3Score(intermediateS3);
		Approvals.verifyAsJson(s3);
	}
	
	public CollectionDocument doc(String docContent, int id) {
		Document doc = new Document();
		doc.add(new StoredField(LuceneDocumentGenerator.FIELD_BODY, docContent));
		doc.add(new StoredField(LuceneDocumentGenerator.FIELD_ID, String.valueOf(id)));

		return CollectionDocument.fromLuceneDocument(doc);
	}
	
	private DocumentHash docHash(String hash, int nGramms) {
		DocumentHash ret = new DocumentHash();
		ret.setFullyCanonicalizedMd5(hash);
		ret.setFullyCanonicalizedWord8GrammSetSize(nGramms);
		
		return ret;
	}
}
