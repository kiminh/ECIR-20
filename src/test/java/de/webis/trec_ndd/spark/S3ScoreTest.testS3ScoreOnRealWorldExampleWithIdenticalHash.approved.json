{
  "commonNGramms": 0,
  "s3Score": 1.0,
  "chunksInA": 0.0,
  "chunksInB": 0.0,
  "meanChunks": 0.0
}